### DISCUSSION TITLE
About LocalFs

### DISCUSSION THREAD

**Discussion by zhangtianxiao** [activej/activej#231] [Q&A] [answered]
In official example, this is ok
```java
ChannelSupplier.of(bytebuf).streamTo(fs.upload(filename))
```
But why
```kotlin
    fs.upload(filename).then { it -> it.accept(bytebuf) }.whenComplete { t, u ->
      fs.download(filename).then { channel, _ ->
        channel.get()
      }.whenComplete { buf, _ ->
        println(ByteBufStrings.asUtf8(buf))
      }
    }
```
file will be upload to `.upload` dir, not `filename`

**Comment by eduard-vasinskyi** [comment, marked as answer]
Hi, @zhangtianxiao

Usually, you would not call `ChannelConsumer#accept` method directly but use the `streamTo()` method instead. This is because there are some invariants that must be respected when working with CSP suppliers and consumers. For example, you only call the `accept()` method after a previous `accept()` call has finished successfully. Or the fact that streaming is considered finished only when `null` value is accepted by the `ChannelConsumer` (indicating the “end-of-stream”). The `streamTo()` method keeps track of all of these invariants, so you should be using `streamTo()` for normal use cases.

In your `LocalActiveFs` example, the file is uploaded to the `.upload` directory. This is a temporary directory for file uploads. `ActiveFs` has an invariant that files are uploaded atomically (either file is fully uploaded or is not uploaded at all). A file will be atomically moved from the `.upload` directory to its “real” location only after streaming is complete. But in your example, you never “finish” streaming data, as you never call `it.accept(null)` or `it.acceptEndOfStream()`.
 
There are several ways that you can fix this:
- Explicitly finish streaming by sending `null` or “end-of-stream”:
```java
fs.upload(filename)
    .then(consumer -> consumer.accept(bytebuf)
        .then(() -> consumer.acceptEndOfStream()))
```
- Use the `streamTo()` method:
```java
fs.upload(filename)
    .then(consumer -> ChannelSupplier.of(byteBuf).streamTo(consumer))
```
